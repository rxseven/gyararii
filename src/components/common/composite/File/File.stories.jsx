import React from 'react';
import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';

import File from './index';

const title = 'Components/Common/Composite/File';
const actions = { onChange: action('onChange') };

storiesOf(title, module)
  .add('default', () => <File {...actions}>Upload</File>, {
    notes: 'Let a user choose one or more files from their device storage.'
  })
  .add(
    'with icon',
    () => (
      <File {...actions} icon="cloud-upload-alt" id="avatar">
        Upload
      </File>
    ),
    {
      notes: {
        markdown: `Pass icon name (string or array of string). Check out all available icons from [Font Awesome](https://fontawesome.com/icons?d=gallery).
        `
      }
    }
  )
  .add(
    'disabled',
    () => (
      <File {...actions} icon="cloud-upload-alt" id="avatar" isLoading>
        Upload
      </File>
    ),
    {
      notes: 'Disable a button while a network request is being submitted.'
    }
  )
  .add(
    'with spinner',
    () => (
      <File {...actions} id="avatar" isLoading spinner>
        Upload
      </File>
    ),
    {
      notes:
        'Disable a button and show a spinner while a network request is being submitted.'
    }
  )
  .add(
    'multiple',
    () => (
      <File {...actions} id="images" multiple>
        Upload multiple files
      </File>
    ),
    {
      notes: 'To allow multiple files to be selected.'
    }
  );
