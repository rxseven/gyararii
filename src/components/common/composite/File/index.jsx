import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';

import Button from 'components/common/base/Button';

const Label = styled.label`
  margin-bottom: 0;
`;

const Input = styled.input`
  &[type='file'] {
    visibility: hidden;
    position: absolute;
  }
`;

const propTypes = {
  children: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  loading: PropTypes.bool,
  multiple: PropTypes.bool,
  onChange: PropTypes.func.isRequired,
  spinner: PropTypes.bool
};

const defaultProps = {
  loading: false,
  multiple: false,
  spinner: false
};

function File({ children, id, loading, multiple, onChange, ...props }) {
  return (
    <Label htmlFor={id}>
      <Button {...props} loading={loading} look="primary" passive>
        {children}
      </Button>
      <Input
        disabled={loading}
        id={id}
        multiple={multiple}
        onChange={onChange}
        type="file"
      />
    </Label>
  );
}

File.propTypes = propTypes;
File.defaultProps = defaultProps;

export default File;
