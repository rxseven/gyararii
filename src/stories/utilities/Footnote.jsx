import styled from 'styled-components';

const Footnote = styled.div`
  bottom: 2rem;
  font-size: 14px;
  position: fixed;
`;

export default Footnote;
