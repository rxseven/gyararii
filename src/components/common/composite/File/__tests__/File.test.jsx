import { fireEvent } from 'react-testing-library';

import { factory } from 'tests/utilities';
import File from '../index';

// Arrange
const seed = { content: 'content' };
const source = {
  children: seed.content,
  id: 'file',
  onUpload: jest.fn()
};
const input = { ...seed, ...source };

// Setup
function setup(props) {
  return factory(File, source, props);
}

// Test suites
describe('<File />', () => {
  it('should render without crashing', () => {
    setup();
  });

  it('should render passed children correctly', () => {
    const expected = { content: input.children };
    const { component } = setup();

    expect(component).toHaveTextContent(expected.content);
  });

  it('should allow a single file to be uploaded by default', () => {
    const expected = { attribute: 'multiple' };
    const { getByLabelText } = setup();

    expect(getByLabelText('upload')).not.toHaveAttribute(expected.attribute);
  });

  it('should accept multiple files to be uploaded', () => {
    const props = { multiple: true };
    const expected = { attribute: 'multiple' };
    const { getByLabelText } = setup(props);

    expect(getByLabelText('upload')).toHaveAttribute(expected.attribute);
  });

  it('should call onUpload() when input field has changed', () => {
    const expected = { called: 1 };
    const { getByLabelText } = setup();

    fireEvent.change(getByLabelText('upload'));

    expect(input.onUpload).toHaveBeenCalledTimes(expected.called);
  });

  it('should disabled input field when loading data', () => {
    const props = { isLoading: true };
    const { getByLabelText } = setup(props);

    expect(getByLabelText('upload')).toBeDisabled();
  });
});
