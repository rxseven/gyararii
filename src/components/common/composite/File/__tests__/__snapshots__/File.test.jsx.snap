// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<File /> Snapshot tests should render correctly 1`] = `
<label
  className="sc-ifAKCX bAELOi"
  htmlFor="file"
>
  <span
    className="btn btn-primary btn-sm"
    disabled={false}
    onChange={[MockFunction]}
    role="button"
  >
    <span
      className="sc-bwzfXH foyPXx"
    >
      Text
    </span>
    <span
      className="sc-bxivhb cRFcgP"
    >
      <span>
        <style>
           @-webkit-keyframes brsk-three-bounce { 0%, 80%, 100% { -webkit-transform: scale(0); } 40% { -webkit-transform: scale(1); } } @keyframes brsk-three-bounce { 0%, 80%, 100% { transform: scale(0); -webkit-transform: scale(0); } 40% { transform: scale(1); -webkit-transform: scale(1); } } 
        </style>
        <div
          style={
            Object {
              "WebkitAnimationDelay": "-0.32s",
              "WebkitAnimationDuration": "1.25s",
              "WebkitAnimationIterationCount": "infinite",
              "WebkitAnimationName": "brsk-three-bounce",
              "WebkitAnimationTimingFunction": "ease-in-out",
              "animationDelay": "-0.32s",
              "animationDuration": "1.25s",
              "animationIterationCount": "infinite",
              "animationName": "brsk-three-bounce",
              "animationTimingFunction": "ease-in-out",
              "backgroundColor": "#fff",
              "borderRadius": "100%",
              "display": "inline-block",
              "height": 7,
              "marginRight": 5,
              "width": 7,
            }
          }
        />
        <div
          style={
            Object {
              "WebkitAnimationDelay": "-0.16s",
              "WebkitAnimationDuration": "1.25s",
              "WebkitAnimationIterationCount": "infinite",
              "WebkitAnimationName": "brsk-three-bounce",
              "WebkitAnimationTimingFunction": "ease-in-out",
              "animationDelay": "-0.16s",
              "animationDuration": "1.25s",
              "animationIterationCount": "infinite",
              "animationName": "brsk-three-bounce",
              "animationTimingFunction": "ease-in-out",
              "backgroundColor": "#fff",
              "borderRadius": "100%",
              "display": "inline-block",
              "height": 7,
              "marginRight": 5,
              "width": 7,
            }
          }
        />
        <div
          style={
            Object {
              "WebkitAnimationDuration": "1.25s",
              "WebkitAnimationIterationCount": "infinite",
              "WebkitAnimationName": "brsk-three-bounce",
              "WebkitAnimationTimingFunction": "ease-in-out",
              "animationDuration": "1.25s",
              "animationIterationCount": "infinite",
              "animationName": "brsk-three-bounce",
              "animationTimingFunction": "ease-in-out",
              "backgroundColor": "#fff",
              "borderRadius": "100%",
              "display": "inline-block",
              "height": 7,
              "width": 7,
            }
          }
        />
      </span>
    </span>
  </span>
  <input
    className="sc-EHOje dQDKDv"
    disabled={false}
    id="file"
    multiple={false}
    type="file"
  />
</label>
`;
